README
-- $MawkId: README,v 1.4 2009/12/20 15:54:32 tom Exp $
-- vile:txtmode fc=78

mawk -- an implementation of new/posix awk
version 1.3.4

Generic installation instructions are in file INSTALL.
This file gives more specific information.

Send bug reports, comments, questions, etc. to 

	Thomas E. Dickey <dickey@invisible-island.net>
	http://invisible-island.net/mawk/
	ftp://invisible-island.net/mawk/

-------------------------------------------------------------------------------

Portability:
-----------

	This program builds/runs on several POSIX-style platforms.
	It has been recently tested for these:

	AIX (5.1 and 5.3, using cc and gcc)
	Cygwin (1.5.21 on Windows/XP)
	FreeBSD (6.0)
	HPUX (10.20, 11.00, 11.11 and 11.23, using cc and gcc)
	IRIX64 (using cc and gcc)
	Linux (several flavors, using gcc and icc)
	MinGW/MSYS (on Windows/XP)
	OpenBSD (4.1)
	Solaris (2.6 to 10, using cc and gcc)
	Tru64 (4.0D and 5.1, using cc and gcc)

Options:
-------

	The configure script has these application-specific options:

	--disable-echo          display "compiling" commands

		Show "compiling foo.c" rather than the full compiler and
		options.  Those are clutter useful only to a developer. 
		Developers focus on compiler warnings anyway.

	--enable-warnings       test: turn on gcc compiler warnings

		This turns on the usual gcc compiler warnings needed to
		do useful development.  If you happen to be using the
		Intel compiler icc, it does the right thing for that.

	--with-builtin-regex    use mawk's own regular-expressions engine

		Normally mawk's configure script uses the built-in regular
		expressions.  The choice of default is based on where the
		greatest lossage occurs.  More scripts use gawk's
		null-character extension than use POSIX brace expressions.  If
		your needs are different, build mawk using an external library.

Limitations:
-----------

	mawk 1.3.3 was developed to correspond with a POSIX draft.  POSIX
	continued, and incorporated some of mawk's extensions as features. 
	The 1.3.4 release fills in the major areas in which POSIX grew past
	the mawk 1.3.3 implementation.

	mawk's built-in regular expression engine does not yet support brace
	expressions, e.g.,

		/a{,4}/

	Use an external regular expression library if you require brace
	expressions.  Aside from that, mawk's built-in regular expressions
	provide POSIX functionality.

	Using an external regular expression library means that mawk cannot
	match expressions containing the null character.  That is a
	nonstandard feature provided by gawk.  Using mawk's built-in regular
	expressions, there is some support for this feature.

	Aside from supporting nulls, using an external regular expression
	library is usually advantageous.

	The Cygwin math library has some problems (which are also visible in
	the gawk port).  For instance, its log() function returns Inf
	(infinity) for log(-8) rather than NaN (not a number) as all of the
	Unix and similar platforms would do.

	The MSYS package for regular expressions is unusable as an external
	library for mawk because it generates a runtime error when asked to
	compile a newline, e.g., "\n".  mawk does this during initialization.

	The MinGW/MSYS port does not currently support pipes.  There is
	source-code from the obsolete MS-DOS port which can be reused for this
	purpose; however that is not the focus of the 1.3.4 release.


CHANGES
-- $MawkId: CHANGES,v 1.136 2010/12/07 13:58:48 tom Exp $

Changes by Thomas E Dickey <dickey@invisible-island.net>

20101210
	+ update copyright notices to reflect changes made starting 20080909.

20101207
	+ add configure --enable-trace, to consolidate various debugging
	  traces which were written to stderr.  This writes to Trace.out.
	+ modify zmalloc.c to make it simpler to use the --disable-leaks
	  configure script option to analyze memory leaks.  The memory pool
	  which interferes with analysis can be disabled by defining
	  DEBUG_ZMALLOC.
	+ (IN-PROGRESS) convert gsub() function from recursive, to
	  non-recursive to lessen use of program stack (Debian #158481).
	+ escape some dashes in manpage, to work with groff.
	+ change section for Debian package to "interpreters" (Debian #442071).
	+ fix printable token value for SC_MATCH in scan.c (report by Heiner
	  Marxen).

20100625
	+ correct translation of octal and hex escapes for system regular
	  expression library.
	+ modify configure script to support --program-suffix, etc.
	+ add Debian package scripts, for "mawk-cur".
	+ add RPM spec-file.
	+ move release- and patch-level values from version.c to patchlev.h
	  to simplify packaging scripts.

20100618
	+ correct translation of "^{" pattern for system regular expression
	  library  (report by Elias Pipping).
	+ fix sentence fragment in README (report by Elias Pipping).

20100507
	+ cleanup gcc warnings for 64-bit platform, e.g., use size_t rather
	  than unsigned, etc.
	+ fix warnings from clang --analyze
	+ update/improve configure script
	  + modify CF_GCC_VERSION to ignore stderr, e.g., from c89 alias
	  + modify CF_GCC_WARNINGS, moving -W and -Wall into the list to check,
	    since c89 alias for gcc complains about these options.
	  + add --disable-leaks and related options, for testing.
	  + add lint rule to makefile.
	  + add configure-check for ctags to work with pkgsrc.
	+ amend change of array.w, fixes a regression in "delete" (report by
	  Heiner Marxen).

20100419
	+ modify split() to handle embedded nulls in the string to split, e.g.,
		BEGIN{s="a\0b"; print length(s); n = split(s,f,""); print n}
	  (report by Morris Lee).

	+ modify array.w to update table pointers in the special case where
	  an array is known to have string-indices, but is later indexed via
	  integers.  The problem occurs when the array grows large enough to
	  rehash it, e.g.,
		BEGIN{a["n"];for(i=1;i<1000;++i)printf "%d\n", a[i]; }
	  (report by Morris Lee).

	+ increase size of reference-count for strings to unsigned.  It was an
	  unsigned short, which prevented using arrays larger than 64k, e.g.,
		BEGIN{for(i = 1; i <= 65550; ++i){if(i >= 65534 && i<= 65537) print i; s[i] = "a"}; delete s;}
	  (report by Morris Lee).

	+ add special case for Solaris 10 (and up) to configure check
	  CF_XOPEN_SOURCE

	+ refactored configure check CF_REGEX

20100224
	+ add a configure check for large files (report by Sean Kumar).

	+ modify check in collect_RE() to show the actual limit value, e.g.,
	  MIN_SPRINTF-2 used for built-in regular expressions.

	+ increase MIN_SPRINTF, used as limit on regular-expression size, to
	  match the MAX_SPLIT value, i.e., slightly more than doubling the size
	  (report by Markus Gnam).

	+ further modify makefile to build outside the source-tree.

	+ modify makefile and mawktest to use relative path again, since the
	  existing script did not work with openSUSE's build (patch by Guido
	  Berhoerster).

	+ fix makefile's .c.i rule, which lacked CPP definition.

	+ update mawktest.bat script to more/less correspond with mawktest,
	  for Win32 console except where echo command does not handle the
	  required quoting syntax.

	+ add vs6.mak, for Visual Studio 6.

	+ modify mawktest script to report results from all tests, rather than
	  halting on the first failure.

	+ add limit-check after processing match(test, "[^0-9A-Za-z]") to
	  ensure the internal trailing null of the test-string is not mistaken
	  for part of the string, i.e., RSTART, RLENGTH are computed correctly
	  (report by Markus Gnam).

	+ modify parsing of -W option to use comma-separated values, e.g.,
	  "-Wi,e" for "-Winteractive" and "-Werror".

	+ add timestamp to scancode.c, to help manage revisions.

	+ improve configure macro CF_XOPEN_SOURCE, making it remove possibly
	  conflicting definitions before adding new ones.

	+ update config.guess and config.sub

	> patches by Jan Psota:

	+ improve buffering for -Winteractive option.

	+ allow multiple single-character flags after -W, e.g., "-Wie" for
	  "-Winteractive" and "-Werror" to permit these to be passed on a
	  "#"-line of a shell script, e.g.,
	  	#!/usr/bin/mawk -Wie

	> patches by Jonathan Nieder:

	+ add new M_SAVE_POS and M_2JC operation codes (states) to the
	  built-in regular expression engine.  Use these to reimplement
	  m* (closure), to provide a way to avoid infinite looping on
	  matches against empty strings.  This change requires reimplementing
	  the workaround for gawk's noloop1 testcase from 20090726.

	+ improve buffer-overflow check for string_buff.

	+ fix collect_RE to treat "[^]]" as a character class (meaning "not a
	  closing bracket") but "[^^]]" not as one.  This also requires
	  initializing the local "start of character class" variable to NULL
	  rather than the beginning of the string, to avoid an invalid array
	  access when collecting expressions such as "^text".

	+ within a character class and not followed by a :, ., or ~, a "[" is
	  just like any other character.  This way, you can tell mawk to scan
	  for a literal [ character with "mawk /[[]/", and you can scan for a [
	  or ] with "mawk /[][]/".  Also clean up the relevant loop in
	  do_class() to make it a bit more readable.

	+ outside a character class, a "]" is just like any other character.

	+ prevent do_class() from scanning past the end, e.g., if the
	  terminating zero byte was escaped.


da.c
/********************************************
da.c
copyright 2008-2009,2010, Thomas E. Dickey
copyright 1991-1994,1995, Michael D. Brennan

This is a source file for mawk, an implementation of
the AWK programming language.

Mawk is distributed without warranty under the terms of
the GNU General Public License, version 2, 1991.
********************************************/

/*
 * $MawkId: da.c,v 1.10 2010/07/31 21:30:31 tom Exp $
 * @Log: da.c,v @
 * Revision 1.6  1995/06/18  19:19:59  mike
 * remove use of comma operator that broke some sysVr3 compilers
 *
 * Revision 1.5  1994/12/13  00:12:08  mike
 * delete A statement to delete all of A at once
 *
 * Revision 1.4  1994/10/08  19:15:32  mike
 * remove SM_DOS
 *
 * Revision 1.3  1993/12/01  14:25:10  mike
 * reentrant array loops
 *
 * Revision 1.2  1993/07/22  00:04:05  mike
 * new op code _LJZ _LJNZ
 *
 * Revision 1.1.1.1  1993/07/03	 18:58:10  mike
 * move source to cvs
 *
 * Revision 5.4	 1993/01/09  19:05:48  mike
 * dump code to stdout and exit(0)
 *
 * Revision 5.3	 1993/01/07  02:50:33  mike
 * relative vs absolute code
 *
 * Revision 5.2	 1992/07/25  21:35:25  brennan
 * patch2
 * fixed small typo on da of _PRE_DEC
 *
 * Revision 5.1	 1991/12/05  07:55:45  brennan
 * 1.1 pre-release
 *
 */

/*  da.c  */
/*  disassemble code */

#include  "mawk.h"

#include  "code.h"
#include  "bi_funct.h"
#include  "repl.h"
#include  "field.h"

typedef struct {
    char op;
    const char *name;
} OP_NAME;

static const char *find_bi_name(PF_CP);
/* *INDENT-OFF* */
static OP_NAME simple_code[] =

{
   {_STOP,      "stop"},
   {FE_PUSHA,   "fe_pusha"},
   {FE_PUSHI,   "fe_pushi"},
   {A_TEST,     "a_test"},
   {A_DEL,      "a_del"},
   {DEL_A,      "del_a"},
   {POP_AL,     "pop_al"},
   {_POP,       "pop"},
   {_ADD,       "add"},
   {_SUB,       "sub"},
   {_MUL,       "mul"},
   {_DIV,       "div"},
   {_MOD,       "mod"},
   {_POW,       "pow"},
   {_NOT,       "not"},
   {_UMINUS,    "uminus"},
   {_UPLUS,     "uplus"},
   {_TEST,      "test"},
   {_CAT,       "cat"},
   {_ASSIGN,    "assign"},
   {_ADD_ASG,   "add_asg"},
   {_SUB_ASG,   "sub_asg"},
   {_MUL_ASG,   "mul_asg"},
   {_DIV_ASG,   "div_asg"},
   {_MOD_ASG,   "mod_asg"},
   {_POW_ASG,   "pow_asg"},
   {NF_PUSHI,   "nf_pushi"},
   {F_ASSIGN,   "f_assign"},
   {F_ADD_ASG,  "f_add_asg"},
   {F_SUB_ASG,  "f_sub_asg"},
   {F_MUL_ASG,  "f_mul_asg"},
   {F_DIV_ASG,  "f_div_asg"},
   {F_MOD_ASG,  "f_mod_asg"},
   {F_POW_ASG,  "f_pow_asg"},
   {_POST_INC,  "post_inc"},
   {_POST_DEC,  "post_dec"},
   {_PRE_INC,   "pre_inc"},
   {_PRE_DEC,   "pre_dec"},
   {F_POST_INC, "f_post_inc"},
   {F_POST_DEC, "f_post_dec"},
   {F_PRE_INC,  "f_pre_inc"},
   {F_PRE_DEC,  "f_pre_dec"},
   {_EQ,        "eq"},
   {_NEQ,       "neq"},
   {_LT,        "lt"},
   {_LTE,       "lte"},
   {_GT,        "gt"},
   {_GTE,       "gte"},
   {_MATCH2,    "match2"},
   {_EXIT,      "exit"},
   {_EXIT0,     "exit0"},
   {_NEXT,      "next"},
   {_RET,       "ret"},
   {_RET0,      "ret0"},
   {_OMAIN,     "omain"},
   {_JMAIN,     "jmain"},
   {OL_GL,      "ol_gl"},
   {OL_GL_NR,   "ol_gl_nr"},
   {_HALT, (char *) 0}
} ;
/* *INDENT-ON* */

static const char *jfmt = "%s%s%03d\n";
   /* format to print jumps */
static const char *tab2 = "\t\t";

void
da_string(FILE *fp, const char *str, size_t len)
{
    size_t n;

    fputc('"', fp);
    for (n = 0; n < len; ++n) {
	UChar ch = (UChar) str[n];
	switch (ch) {
	case '\\':
	    fprintf(fp, "\\\\");
	    break;
	case '"':
	    fprintf(fp, "\"");
	    break;
	default:
	    if (ch >= 32 && ch < 127)
		fprintf(fp, "%c", ch);
	    else
		fprintf(fp, "\\%03o", ch);
	    break;
	}
    }
    fputc('"', fp);
}

void
da(INST * start, FILE *fp)
{
    CELL *cp;
    register INST *p = start;
    const char *name;

    while (p->op != _HALT) {
	/* print the relative code address (label) */
	fprintf(fp, "%03ld ", (long) (p - start));

	switch (p++->op) {

	case _PUSHC:
	    cp = (CELL *) p++->ptr;
	    switch (cp->type) {
	    case C_RE:
		fprintf(fp, "pushc\t0x%lx\t/%s/\n", (long) cp->ptr,
			re_uncompile(cp->ptr));
		break;

	    case C_SPACE:
		fprintf(fp, "pushc\tspace split\n");
		break;

	    case C_SNULL:
		fprintf(fp, "pushc\tnull split\n");
		break;
	    case C_REPL:
		fprintf(fp, "pushc\trepl\t%s\n",
			repl_uncompile(cp));
		break;
	    case C_REPLV:
		fprintf(fp, "pushc\treplv\t%s\n",
			repl_uncompile(cp));
		break;

	    default:
		fprintf(fp, "pushc\tWEIRD\n");;
		break;
	    }
	    break;

	case _PUSHD:
	    fprintf(fp, "pushd\t%.6g\n", *(double *) p++->ptr);
	    break;
	case _PUSHS:
	    {
		STRING *sval = (STRING *) p++->ptr;
		fprintf(fp, "pushs\t");
		da_string(fp, sval->str, sval->len);
		fprintf(fp, "\n");
		break;
	    }

	case _MATCH0:
	case _MATCH1:
	    fprintf(fp, "match%d\t0x%lx\t/%s/\n",
		    p[-1].op == _MATCH1, (long) p->ptr,
		    re_uncompile(p->ptr));
	    p++;
	    break;

	case _PUSHA:
	    fprintf(fp, "pusha\t%s\n",
		    reverse_find(ST_VAR, &p++->ptr));
	    break;

	case _PUSHI:
	    cp = (CELL *) p++->ptr;
	    if (cp == field)
		fprintf(fp, "pushi\t$0\n");
	    else if (cp == &fs_shadow)
		fprintf(fp, "pushi\t@fs_shadow\n");
	    else {
		if (
#ifdef  MSDOS
		       SAMESEG(cp, field) &&
#endif
		       cp > NF && cp <= LAST_PFIELD)
		    name = reverse_find(ST_FIELD, &cp);
		else
		    name = reverse_find(ST_VAR, &cp);

		fprintf(fp, "pushi\t%s\n", name);
	    }
	    break;

	case L_PUSHA:
	    fprintf(fp, "l_pusha\t%d\n", p++->op);
	    break;

	case L_PUSHI:
	    fprintf(fp, "l_pushi\t%d\n", p++->op);
	    break;

	case LAE_PUSHI:
	    fprintf(fp, "lae_pushi\t%d\n", p++->op);
	    break;

	case LAE_PUSHA:
	    fprintf(fp, "lae_pusha\t%d\n", p++->op);
	    break;

	case LA_PUSHA:
	    fprintf(fp, "la_pusha\t%d\n", p++->op);
	    break;

	case F_PUSHA:
	    cp = (CELL *) p++->ptr;
	    if (
#ifdef  MSDOS
		   SAMESEG(cp, field) &&
#endif
		   cp >= NF && cp <= LAST_PFIELD)
		fprintf(fp, "f_pusha\t%s\n",
			reverse_find(ST_FIELD, &cp));
	    else
		fprintf(fp, "f_pusha\t$%d\n",
			field_addr_to_index(cp));
	    break;

	case F_PUSHI:
	    p++;
	    fprintf(fp, "f_pushi\t$%d\n", p++->op);
	    break;

	case AE_PUSHA:
	    fprintf(fp, "ae_pusha\t%s\n",
		    reverse_find(ST_ARRAY, &p++->ptr));
	    break;

	case AE_PUSHI:
	    fprintf(fp, "ae_pushi\t%s\n",
		    reverse_find(ST_ARRAY, &p++->ptr));
	    break;

	case A_PUSHA:
	    fprintf(fp, "a_pusha\t%s\n",
		    reverse_find(ST_ARRAY, &p++->ptr));
	    break;

	case _PUSHINT:
	    fprintf(fp, "pushint\t%d\n", p++->op);
	    break;

	case _BUILTIN:
	    fprintf(fp, "%s\n",
		    find_bi_name((PF_CP) p++->ptr));
	    break;

	case _PRINT:
	    fprintf(fp, "%s\n",
		    (PF_CP) p++->ptr == bi_printf
		    ? "printf" : "print");
	    break;

	case _JMP:
	    fprintf(fp, jfmt, "jmp", tab2, (p - start) + p->op);
	    p++;
	    break;

	case _JNZ:
	    fprintf(fp, jfmt, "jnz", tab2, (p - start) + p->op);
	    p++;
	    break;

	case _JZ:
	    fprintf(fp, jfmt, "jz", tab2, (p - start) + p->op);
	    p++;
	    break;

	case _LJZ:
	    fprintf(fp, jfmt, "ljz", tab2, (p - start) + p->op);
	    p++;
	    break;

	case _LJNZ:
	    fprintf(fp, jfmt, "ljnz", tab2 + 1, (p - start) + p->op);
	    p++;
	    break;

	case SET_ALOOP:
	    fprintf(fp, "set_al\t%03ld\n", (long) (p + p->op - start));
	    p++;
	    break;

	case ALOOP:
	    fprintf(fp, "aloop\t%03ld\n", (long) (p - start + p->op));
	    p++;
	    break;

	case A_CAT:
	    fprintf(fp, "a_cat\t%d\n", p++->op);
	    break;

	case _CALL:
	    fprintf(fp, "call\t%s\t%d\n",
		    ((FBLOCK *) p->ptr)->name, p[1].op);
	    p += 2;
	    break;

	case _RANGE:
	    fprintf(fp, "range\t%03ld %03ld %03ld\n",
	    /* label for pat2, action, follow */
		    (long) (p - start + p[1].op),
		    (long) (p - start + p[2].op),
		    (long) (p - start + p[3].op));
	    p += 4;
	    break;
	default:
	    {
		OP_NAME *q = simple_code;
		int k = (p - 1)->op;

		while (q->op != _HALT && q->op != k)
		    q++;

		fprintf(fp, "%s\n",
			q->op != _HALT ? q->name : "bad instruction");
	    }
	    break;
	}
    }
    fflush(fp);
}
/* *INDENT-OFF* */
static struct
{
   PF_CP action ;
   const char *name ;
}
special_cases[] =
{
   {bi_split,   "split"},
   {bi_match,   "match"},
   {bi_getline, "getline"},
   {bi_sub,     "sub"},
   {bi_gsub,    "gsub"},
   {(PF_CP) 0, (char *) 0}
} ;
/* *INDENT-ON* */

static const char *
find_bi_name(PF_CP p)
{
    BI_REC *q;
    int i;

    for (q = bi_funct; q->name; q++) {
	if (q->fp == p) {
	    /* found */
	    return q->name;
	}
    }
    /* next check some special cases */
    for (i = 0; special_cases[i].action; i++) {
	if (special_cases[i].action == p) {
	    return special_cases[i].name;
	}
    }

    return "unknown builtin";
}

static struct fdump {
    struct fdump *link;
    FBLOCK *fbp;
} *fdump_list;			/* linked list of all user functions */

void
add_to_fdump_list(FBLOCK * fbp)
{
    struct fdump *p = ZMALLOC(struct fdump);
    p->fbp = fbp;
    p->link = fdump_list;
    fdump_list = p;
}

void
fdump(void)
{
    register struct fdump *p, *q = fdump_list;

    while (q) {
	p = q;
	q = p->link;
	fprintf(stdout, "function %s\n", p->fbp->name);
	da(p->fbp->code, stdout);
	ZFREE(p);
    }
}

#ifdef NO_LEAKS
/* *INDENT-OFF* */
static OP_NAME other_codes[] = {
    { AE_PUSHA,   "ae_pusha" },
    { AE_PUSHI,   "ae_pushi" },
    { ALOOP,      "aloop" },
    { A_CAT,      "a_cat" },
    { A_PUSHA,    "a_pusha" },
    { F_PUSHA,    "f_pusha" },
    { F_PUSHI,    "f_pushi" },
    { LAE_PUSHA,  "lae_pusha" },
    { LAE_PUSHI,  "lae_pushi" },
    { LA_PUSHA,   "la_pusha" },
    { L_PUSHA,    "l_pusha" },
    { L_PUSHI,    "l_pushi" },
    { SET_ALOOP,  "set_al" },
    { _CALL,      "call" },
    { _JMP,       "jmp" },
    { _JNZ,       "jnz" },
    { _JZ,        "jz" },
    { _LJNZ,      "ljnz" },
    { _LJZ,       "ljz" },
    { _MATCH0,    "match0" },
    { _MATCH1,    "match1" },
    { _PUSHA,     "pusha" },
    { _PUSHC,     "pushc" },
    { _PUSHD,     "pushd" },
    { _PUSHI,     "pushi" },
    { _PUSHINT,   "pushint" },
    { _PUSHS,     "pushs" },
    { _RANGE,     "range" },
    { _HALT,      0 },
};
/* *INDENT-ON* */

const char *
da_op_name(INST * cdp)
{
    int n;
    const char *result = 0;

    if (cdp->op == _BUILTIN) {
	result = find_bi_name((PF_CP) cdp[1].ptr);
    } else if (cdp->op == _PRINT) {
	result = ((PF_CP) cdp->ptr == bi_printf
		  ? "printf"
		  : "print");
    } else if (cdp->op == _HALT) {
	result = "halt";
    } else {
	for (n = 0; simple_code[n].op != _HALT; ++n) {
	    if (simple_code[n].op == cdp->op) {
		result = simple_code[n].name;
		break;
	    }
	}
	if (result == 0) {
	    for (n = 0; other_codes[n].op != _HALT; ++n) {
		if (other_codes[n].op == cdp->op) {
		    result = other_codes[n].name;
		    break;
		}
	    }
	}
	if (result == 0) {
	    result = "unknown";
	}
    }
    return result;
}
#endif
